### ISSUE TITLE
Jobs that kill PHP left as "running"

### ISSUE THREAD

**Issue by sminnee** [symbiote/silverstripe-queuedjobs#152]
When a job triggers a PHP fatal error, the job status is left as "running".

At a minimum, this occurs with a PHP parse error, if the reader struggles to find a replication case.

Ideally this could be cleaned up somehow. An `register_shutdown` function might be able to trap some errors but I'm not sure if it can grab everything, although [this](https://stackoverflow.com/questions/29735022/how-to-handle-parse-error-using-register-shutdown-function-in-php) suggests it can pick up parser errors, which is promising.

Expected behaviour would be that jobs that die mid-way are marked with an error.

**Comment by robbieaverill** [comment]
Another option might be to track the last activity of a job's process, so jobs that haven't performed any activity for a certain period of time can be marked as abandoned, or something.

I like the register_shutdown idea, but if the job dies because of a network issue between PHP and the database it's unlikely to help.

**Comment by sminnee** [comment]
I would probably want to see both in place.

A garbage collection process might be appropriate if there's no activity for an hour or 2 (as long as 73135e676e6e7154e6f35712c82831efa257ace7 is in place), but lessening the time in the default implementation would be dangerous if you end up killing long-running, silent jobs. So it's going to be an eventual tidy-up for rarer cases like DB-network issues.

**Comment by robbieaverill** [comment]
Yeah - I suppose it's more likely that you run a job with a syntax error in some class and it accidentally spawns and kills a bunch of jobs

**Comment by sminnee** [comment]
Recoded as bug, because it's currently broken.

**Comment by NightJar** [comment]
There are a couple of ways to approach this, after chatting with @robbieaverill it sounds like perhaps the most robust would be to register the running PID when the process begins, and garbage collect on shutdown... however this will rely on the POSIX library - which would preclude the modules usage on Windows servers. However I'm not sure about the current support level for Windows - could you please confirm if this is a thing @nyeholt ?

Another approach may be to simply have users add a ` || run_clean-up_task` to the end of the cron job, which would (in theory) catch errors and clean them up. PHP is a little inconsistent in this manner though unfortunately - and again it's not really portable.

**Comment by robbieaverill** [comment]
>An register_shutdown function might be able to trap some errors but I'm not sure if it can grab everything, although [this](https://stackoverflow.com/questions/29735022/how-to-handle-parse-error-using-register-shutdown-function-in-php) suggests it can pick up parser errors, which is promising.

Quick look - it looks like we'd need to register the error handler outside of the SilverStripe bootstrap process, which might mean adding a PHP wrapper file around `vendor/bin/sake dev/tasks...` in order to define and handle parse errors that are caused from within the SilverStripe framework.

Would it be fair to say that while the SilverStripe class manifest exists, all PHP code within it would be parsed at manifest compile time? If so, we're unlikely to see a parse error that doesn't bring down `sake` as well?

Perhaps fatal errors are a more suitable target for this, i.e. method does not exist etc. Our shutdown handler could write the contents of the last error back into the logs for the job (assuming the job ID is still available in the scope).

**Comment by dhensby** [comment]
On this note, I've wanted to register an "early error handler" which is replaced by the SS one once SS boots, because at the moment any early stage errors result in a 200 HTTP response which is not so good... I had worked on something to do this a while ago but never got it to PR state.

https://github.com/silverstripe/silverstripe-framework/commit/97ab0090ec5a5f3e252055074f9149a62796e110

**Comment by robbieaverill** [comment]
We spoke about doing this with a monitoring job to mark things as abandoned etc, but we'd need to be careful of marking jobs that take ages to run without activity as abandoned (perhaps they should be reporting back though...). With tracking process IDs, this can be done in a unix/Windows compliant way but it would be hard to do in an environment with multiple servers running the queue.

**Comment by robbieaverill** [comment]
@ScopeyNZ's suggestion for improving the way we detect stalled jobs in another issue was to start tracking the last activity timestamp and measure a period of time after that with no activity as stalled. If we implement that, it would also fix this since the job would be detected as stalled after a while.

**Comment by GuySartorelli** [comment]
This is an old issue, so just throwing in the context that the issue as described in the original description still occurs in v4.8.0
